src/ButteredBread.sol

```diff
@@ -0,0 +1,177 @@
+ */
+contract ButteredBread is ERC20VotesUpgradeable, OwnableUpgradeable, IButteredBread {
+    uint256 public constant FIXED_POINT_PERCENT = 100;
+
+    /// @notice Access control for Breadchain sanctioned liquidity pools
+    mapping(address lp => bool allowed) public allowlistedLPs;
+    /// @notice How much ButteredBread should be minted for a Liquidity Pool token (Butter)
+    mapping(address lp => uint256 factor) public scalingFactors;
+    /// @notice Butter balance by account and Liquidity Pool token deposited
+    mapping(address account => mapping(address lp => LPData)) internal _accountToLPData;
+
+    modifier onlyAllowed(address _lp) {
+        if (allowlistedLPs[_lp] != true) revert NotAllowListed();
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @param _initData See IButteredBread
+    function initialize(InitData calldata _initData) external initializer {
+        if (_initData.liquidityPools.length != _initData.scalingFactors.length) revert InvalidValue();
+
+        __Ownable_init(msg.sender);
+        __ERC20_init(_initData.name, _initData.symbol);
+
+        for (uint256 i; i < _initData.liquidityPools.length; ++i) {
+            allowlistedLPs[_initData.liquidityPools[i]] = true;
+            scalingFactors[_initData.liquidityPools[i]] = _initData.scalingFactors[i];
+        }
+    }
+
+    /**
+     * @notice The amount of LP tokens (Butter) deposited for a an account
+     * @param _account Voting account
+     * @param _lp Liquidity Pool token
+     */
+    function accountToLPData(address _account, address _lp) external view returns (LPData memory _lpData) {
+        _lpData = _accountToLPData[_account][_lp];
+    }
+
+    /**
+     * @notice Deposit LP tokens
+     * @param _lp Liquidity Pool token
+     * @param _amount Value of LP token
+     */
+    function deposit(address _lp, uint256 _amount) external onlyAllowed(_lp) {
+        _deposit(msg.sender, _lp, _amount);
+    }
+
+    /**
+     * @notice Withdraw LP tokens
+     * @param _lp Liquidity Pool token
+     * @param _amount Value of LP token
+     */
+    function withdraw(address _lp, uint256 _amount) external onlyAllowed(_lp) {
+        _withdraw(msg.sender, _lp, _amount);
+    }
+
+    /**
+     * @notice Allow or deny LP token
+     * @param _lp Liquidity Pool token
+     * @param _allowed Sanction status of LP token
+     */
+    function modifyAllowList(address _lp, bool _allowed) external onlyOwner {
+        allowlistedLPs[_lp] = _allowed;
+    }
+
+    /**
+     * @notice Set LP token scaling factor
+     * @param _lp Liquidity Pool token
+     * @param _factor Scaling percentage incentive of LP token (e.g. 100 = 1X, 150 = 1.5X, 1000 = 10X)
+     */
+    function modifyScalingFactor(address _lp, uint256 _factor) external onlyOwner onlyAllowed(_lp) {
+        if (_factor < 100) revert InvalidValue();
+        scalingFactors[_lp] = _factor;
+    }
```

Comment:
Can we make `syncVotingWeights` and `syncVotingWeight` internal, and add an address array to this method that gets passed to the `syncVotingWeights` method? This way an administrator is less likely to make a mistake and update the scaling factor without the holders

Reply:
there is an array now. left `syncVotingWeights` as external as a means to manually update the list in case of DDOS, but otherwise can be automated